### DISCUSSION TITLE
new_from_image exit silently in cpp version example code windows 10

### DISCUSSION THREAD

**Discussion by zhusihan-python** [libvips/libvips#3196] [Q&A] [answered]
## Questions
**Describe the bug**
build and run the [example](https://www.libvips.org/API/8.11/cpp/) code as below,
```
/* compile with:
 *      g++ -g -Wall example.cc `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

using namespace vips;

int
main (int argc, char **argv)
{ 
  if (VIPS_INIT (argv[0])) 
    vips_error_exit (NULL);

  if (argc != 3)
    vips_error_exit ("usage: %s input-file output-file", argv[0]);

  VImage in = VImage::new_from_file (argv[1],
    VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));

  double avg = in.avg ();

  printf ("avg = %g\n", avg);
  printf ("width = %d\n", in.width ());

  in = VImage::new_from_file (argv[1],
    VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));

  VImage out = in.embed (10, 10, 1000, 1000,
    VImage::option ()->
      set ("extend", "background")->
      set ("background", 128));

  out.write_to_file (argv[2]);

  vips_shutdown ();

  return 0;
}
```
it exit silently in this line
```
  VImage in = VImage::new_from_file (argv[1],
    VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
```
**To Reproduce**
1. run `pkg-config vips-cpp --cflags --libs`
got config as below
`-mms-bitfields -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/libgsf-1 -ID:/softwares/vips-dev-8.13/include/libxml2 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/ -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/librsvg-2.0 -ID:/softwares/vips-dev-8.13/include/gdk-pixbuf-2.0 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/cairo -ID:/softwares/vips-dev-8.13/include/pango-1.0 -ID:/softwares/vips-dev-8.13/include/fribidi -ID:/softwares/vips-dev-8.13/include/cairo -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/freetype2 -ID:/softwares/vips-dev-8.13/include/harfbuzz -ID:/softwares/vips-dev-8.13/include/pixman-1 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/OpenEXR -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/Imath -ID:/softwares/vips-dev-8.13/include/openjpeg-2.5 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/libpng16 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/orc-0.4 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/nifti -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/glib-2.0 -ID:/softwares/vips-dev-8.13/lib/glib-2.0/include -LD:/softwares/vips-dev-8.13/lib -lvips-cpp -lvips -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl`
2.run
`g++ -g -Wall example.cc -o example.exe -mms-bitfields -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/libgsf-1 -ID:/softwares/vips-dev-8.13/include/libxml2 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/ -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/librsvg-2.0 -ID:/softwares/vips-dev-8.13/include/gdk-pixbuf-2.0 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/cairo -ID:/softwares/vips-dev-8.13/include/pango-1.0 -ID:/softwares/vips-dev-8.13/include/fribidi -ID:/softwares/vips-dev-8.13/include/cairo -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/freetype2 -ID:/softwares/vips-dev-8.13/include/harfbuzz -ID:/softwares/vips-dev-8.13/include/pixman-1 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/OpenEXR -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/Imath -ID:/softwares/vips-dev-8.13/include/openjpeg-2.5 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/libpng16 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/orc-0.4 -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/nifti -ID:/softwares/vips-dev-8.13/include -ID:/softwares/vips-dev-8.13/include/glib-2.0 -ID:/softwares/vips-dev-8.13/lib/glib-2.0/include -LD:/softwares/vips-dev-8.13/lib -lvips-cpp -lvips -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl`
got example.exe
3. run
`example.exe aaa.jpg aaa.png`
nothing happened, avg and width printf log did not show, no errors
4.use pyvips to open aaa.jpg, it works fine
![1669616657897](https://user-images.githubusercontent.com/23504604/204208389-ccedc6f9-ab9b-418c-b759-74b4e80a2485.jpg)

**Expected behavior**
open image successful as pyvips

**Actual behavior**
exit without errors, the image is not loaded at all, i try to replace the filename by absolute path, it didn't work too, seems not related to the image file, but something wrong inside new_from_file

**Screenshots**

**Environment**
 - OS: [windows 10]
 - Vips: [vips-dev-8.13]
 - g++: [g++ (GCC) 11.3.0]
 - pyvips: [2.2.1]

**Additional context**

**Comment by jcupitt** [comment, marked as answer]
Hi @zhusihan-python,

I would guess the problem is mixed C++ compilers.  C++ does not define an ABI, so you can't mix compilers within one project. We build the windows libvips zip on debian bullseye, which is on g++10, and you're using g++11.

I would try rebuilding the C++ API with g++11 on your PC. It should be easy -- it only depends on the libvips DLL (C does have an ABI, so you don't need to match compiler there) and libstd. 

You could also downgrade to g++10, though you'd probably need to match the exact compiler version to be safe:

https://packages.debian.org/bullseye/g++
